File: backend-fastapi/tests/api/endpoints/test_users.py
from http import HTTPStatus

import pytest
from conftest import Helpers

from lib.utils import get_image_path


@pytest.mark.asyncio
async def test_fetch_users(helpers: Helpers):
    headers = dict(Authorization=helpers.user_token)
    response = await helpers.client.get("/api/users/", headers=headers)
    assert response.status_code == HTTPStatus.OK
    data = response.json()
    assert "page" in data and data["page"] == 1
    assert "totalPages" in data
    assert "totalCount" in data
    assert "data" in data


@pytest.mark.asyncio
async def test_query_users(helpers: Helpers):
    headers = dict(Authorization=helpers.user_token)
    payload = dict(email=["regex:@gmail.com"], fields=["email", "name"])
    response = await helpers.client.post(
        "/api/users/query", json=payload, headers=headers
    )
    assert response.status_code == HTTPStatus.OK
    data = response.json()
    assert data["page"] == 1
    assert "totalPages" in data
    assert data["totalCount"] == 1
    fetched = data["data"][0]
    assert fetched == {"name": "Slim Beji", "email": "mslimbeji@gmail.com"}


@pytest.mark.asyncio
async def test_create_user_as_admin(helpers: Helpers):
    files = dict(image=open(get_image_path("avatar1.jpg"), "rb"))
    headers = dict(Authorization=helpers.admin_token)
    data = dict(
        name="Test Van Test",
        email="test@test.com",
        password="very_secret",
        isAdmin=True,
    )
    response = await helpers.client.post(
        "/api/users/", data=data, files=files, headers=headers
    )
    assert response.status_code == HTTPStatus.OK
    data = response.json()
    assert data["email"] == "test@test.com"
    assert data["name"] == "Test Van Test"
    assert data["isAdmin"] is True


@pytest.mark.asyncio
async def test_create_user_as_non_admin(helpers: Helpers):
    files = dict(image=open(get_image_path("avatar2.jpg"), "rb"))
    headers = dict(Authorization=helpers.user_token)
    data = dict(
        name="Test Van Test II",
        email="test_2@test.com",
        password="very_secret",
        isAdmin=True,
    )
    response = await helpers.client.post(
        "/api/users/", data=data, files=files, headers=headers
    )
    assert response.status_code == HTTPStatus.UNAUTHORIZED


@pytest.mark.asyncio
async def test_get_user_by_id(helpers: Helpers):
    headers = dict(Authorization=helpers.user_token)
    response = await helpers.client.get(
        f"/api/users/{helpers.user.id}", headers=headers
    )
    assert response.status_code == HTTPStatus.OK
    data = response.json()
    assert data["email"] == "beji.slim@yahoo.fr"
    assert data["name"] == "Mohamed Slim Beji"


@pytest.mark.asyncio
async def test_update_user(helpers: Helpers):
    data = dict(name="Slim El Beji")
    headers = dict(Authorization=helpers.user_token)
    response = await helpers.client.put(
        f"/api/users/{helpers.user.id}", json=data, headers=headers
    )
    assert response.status_code == HTTPStatus.OK
    data = response.json()
    assert data["email"] == "beji.slim@yahoo.fr"
    assert data["name"] == "Slim El Beji"


@pytest.mark.asyncio
async def test_users_cannot_update_others(helpers: Helpers):
    data = dict(name="Slim El Beji")
    headers = dict(Authorization=helpers.user_token)
    response = await helpers.client.put(
        f"/api/users/{helpers.admin.id}", json=data, headers=headers
    )
    assert response.status_code == HTTPStatus.UNAUTHORIZED


@pytest.mark.asyncio
async def test_delete_user_as_non_admin(helpers: Helpers):
    headers = dict(Authorization=helpers.user_token)
    response = await helpers.client.delete(
        f"/api/users/{helpers.user.id}", headers=headers
    )
    assert response.status_code == HTTPStatus.UNAUTHORIZED


@pytest.mark.asyncio
async def test_delete_user_as_admin(helpers: Helpers):
    headers = dict(Authorization=helpers.admin_token)
    response = await helpers.client.delete(
        f"/api/users/{helpers.user.id}", headers=headers
    )
    assert response.status_code == HTTPStatus.OK
    assert response.json()["message"] == f"Deleted user {helpers.user.id}"

File: backend-axum/src/tests/api/test_users.rs
use axum::{
    body::Body,
    http::{Request, StatusCode},
};
use serde_json::{Value, json};
use tower::ServiceExt;

use crate::{lib_::types_::PaginatedData, tests::api::utils::get_simple_user};
use crate::{models::schemas::UserRead, tests::api::utils::get_admin_user};
use crate::{
    static_,
    tests::api::utils::{MultipartTestRequest, get_content_type, parse_json, setup},
};

#[tokio::test]
async fn test_fetch_users() {
    let (app, state) = setup().await;
    let (_, token) = get_simple_user(state).await;

    let response = app
        .oneshot(
            Request::builder()
                .method("GET")
                .uri("/api/users/")
                .header("Authorization", token)
                .body(Body::empty())
                .expect("test error"),
        )
        .await
        .expect("test error");

    assert_eq!(response.status(), StatusCode::OK);
    assert!(get_content_type(&response).contains("application/json"));

    let json: PaginatedData<Value> =
        serde_json::from_value(parse_json(response).await).expect("test error");
    assert_eq!(json.page, 1);
    assert_eq!(json.total_pages, 1);
    assert_eq!(json.total_count, 2);
    assert!(!json.data.is_empty(), "missing data");
}

#[tokio::test]
async fn test_query_users() {
    let (app, state) = setup().await;
    let (_, token) = get_simple_user(state).await;

    let payload = json!({
        "email": ["ilike:@gmail.com"],
        "fields": ["email", "name"]
    });

    let response = app
        .oneshot(
            Request::builder()
                .method("POST")
                .uri("/api/users/search")
                .header("Authorization", token)
                .header("Content-Type", "application/json")
                .body(Body::from(payload.to_string()))
                .expect("test error"),
        )
        .await
        .expect("test error");

    assert_eq!(response.status(), StatusCode::OK);
    assert!(get_content_type(&response).contains("application/json"));

    let json: PaginatedData<Value> =
        serde_json::from_value(parse_json(response).await).expect("test error");
    assert_eq!(json.page, 1);
    assert_eq!(json.total_pages, 1);
    assert_eq!(json.total_count, 1);
    assert_eq!(json.data[0]["name"], "Slim Beji");
    assert_eq!(json.data[0]["email"], "mslimbeji@gmail.com");
}

#[tokio::test]
async fn test_create_user_as_admin() {
    let (app, state) = setup().await;
    let (_, token) = get_admin_user(state).await;

    let (content_type, body_bytes) = MultipartTestRequest::new()
        .add_field("name", "Test Van Test")
        .add_field("email", "test@test.com")
        .add_field("password", "very_secret")
        .add_field("is_admin", "true")
        .add_file("image", &static_::get_image_path("avatar1.jpg"))
        .build();

    let response = app
        .oneshot(
            Request::builder()
                .method("POST")
                .uri("/api/users/")
                .header("Authorization", token)
                .header("Content-Type", content_type)
                .body(Body::from(body_bytes))
                .expect("test error"),
        )
        .await
        .expect("test error");

    assert_eq!(response.status(), StatusCode::OK);
    assert!(get_content_type(&response).contains("application/json"));

    let resp: UserRead = serde_json::from_value(parse_json(response).await).expect("test error");
    assert_eq!(resp.email, "test@test.com");
    assert_eq!(resp.name, "Test Van Test");
    assert!(resp.is_admin);
}

#[tokio::test]
async fn test_create_user_as_non_admin() {
    let (app, state) = setup().await;
    let (_, token) = get_simple_user(state).await;

    let (content_type, body_bytes) = MultipartTestRequest::new()
        .add_field("name", "Test Van Test II")
        .add_field("email", "test_2@test.com")
        .add_field("password", "very_secret")
        .add_field("is_admin", "true")
        .add_file("image", &static_::get_image_path("avatar1.jpg"))
        .build();

    let response = app
        .oneshot(
            Request::builder()
                .method("POST")
                .uri("/api/users/")
                .header("Authorization", token)
                .header("Content-Type", content_type)
                .body(Body::from(body_bytes))
                .expect("test error"),
        )
        .await
        .expect("test error");

    assert_eq!(response.status(), StatusCode::UNAUTHORIZED);
}

#[tokio::test]
async fn test_get_user_by_id() {
    let (app, state) = setup().await;
    let (user, token) = get_simple_user(state).await;

    let response = app
        .oneshot(
            Request::builder()
                .method("GET")
                .uri(format!("/api/users/{}", user.id))
                .header("Authorization", token)
                .body(Body::empty())
                .expect("test error"),
        )
        .await
        .expect("test error");

    assert_eq!(response.status(), StatusCode::OK);
    assert!(get_content_type(&response).contains("application/json"));

    let resp: UserRead = serde_json::from_value(parse_json(response).await).expect("test error");
    assert_eq!(resp.email, "beji.slim@yahoo.fr");
    assert_eq!(resp.name, "Mohamed Slim Beji");
}

#[tokio::test]
async fn test_update_user() {
    let (app, state) = setup().await;
    let (user, token) = get_simple_user(state).await;

    let payload = json!({"name": "Slim El Beji"});

    let response = app
        .oneshot(
            Request::builder()
                .method("PUT")
                .uri(format!("/api/users/{}", user.id))
                .header("Authorization", token)
                .header("Content-Type", "application/json")
                .body(Body::from(payload.to_string()))
                .expect("test error"),
        )
        .await
        .expect("test error");

    assert_eq!(response.status(), StatusCode::OK);
    assert!(get_content_type(&response).contains("application/json"));

    let resp: UserRead = serde_json::from_value(parse_json(response).await).expect("test error");
    assert_eq!(resp.email, "beji.slim@yahoo.fr");
    assert_eq!(resp.name, "Slim El Beji");
}

#[tokio::test]
async fn test_update_other_users() {
    let (app, state) = setup().await;
    let (_, token) = get_simple_user(state.clone()).await;
    let (admin, _) = get_admin_user(state).await;

    let payload = json!({"name": "Slim El Beji"});

    let response = app
        .oneshot(
            Request::builder()
                .method("PUT")
                .uri(format!("/api/users/{}", admin.id))
                .header("Authorization", token)
                .header("Content-Type", "application/json")
                .body(Body::from(payload.to_string()))
                .expect("test error"),
        )
        .await
        .expect("test error");

    assert_eq!(response.status(), StatusCode::UNAUTHORIZED);
    assert!(get_content_type(&response).contains("application/json"));
}

#[tokio::test]
async fn test_delete_user_as_admin() {
    let (app, state) = setup().await;
    let (admin, token) = get_admin_user(state).await;

    let response = app
        .oneshot(
            Request::builder()
                .method("DELETE")
                .uri(format!("/api/users/{}", admin.id))
                .header("Authorization", token)
                .body(Body::empty())
                .expect("test error"),
        )
        .await
        .expect("test error");

    assert_eq!(response.status(), StatusCode::OK);
    assert!(get_content_type(&response).contains("application/json"));

    let json = parse_json(response).await;
    assert_eq!(json["message"], format!("Deleted user {}", admin.id));
}

#[tokio::test]
async fn test_delete_user_as_non_admin() {
    let (app, state) = setup().await;
    let (user, token) = get_simple_user(state).await;

    let response = app
        .oneshot(
            Request::builder()
                .method("DELETE")
                .uri(format!("/api/users/{}", user.id))
                .header("Authorization", token)
                .body(Body::empty())
                .expect("test error"),
        )
        .await
        .expect("test error");

    assert_eq!(response.status(), StatusCode::UNAUTHORIZED);
    assert!(get_content_type(&response).contains("application/json"));
}
